mne/beamformer/_rap_music.py

```diff
@@ -0,0 +1,239 @@
+        Forward operator.
+    noise_cov : Covariance
+        The noise covariance.
+    label : Label
+        Restricts the rap-music solution to a given label.
+        XXX: not implemented yet.
+    r: int
+        The dimension of the subspace spanning the signal.
+        The default value is 15.
+    n_sources: int
+        The number of sources to estimate.
+    picks : array-like of int | None
+        Indices (in info) of data channels. If None, MEG and EEG data channels
+        (without bad channels) will be used.
+    pick_ori : None | 'normal' | 'max-power'
+        If 'normal', rather than pooling the orientations by taking the norm,
+        only the radial component is kept. If 'max-power', the source
+        orientation that maximizes output source power is chosen.
+        XXX: 'max-power' not implemented
+    return_residual : bool
+        If True, the residual is returned as an Evoked instance.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+    Returns
+    -------
+    stc : SourceEstimate
+        Source time courses.
+    D : array
+        Data explained by the sources. Computed only if return_residual
+        is True.
+    """
+    is_free_ori, picks, ch_names, proj, vertno, G =\
+        _prepare_beamformer_input(info, forward, label, picks, pick_ori)
+
+    # Handle whitening + data covariance
+    whitener, _ = compute_whitener(noise_cov, info, picks)
+
+    # whiten the leadfield
+    G = np.dot(whitener, G)
+    data = np.dot(whitener, data)
+
+    # Pick source orientation normal to cortical surface
+    if pick_ori == 'normal':
+        G = G[:, 2::3]
+        is_free_ori = False
+
+    eig_values, eig_vectors = linalg.eigh(np.dot(data, data.T))
+    phi_sig = eig_vectors[:, -r:]
+
+    n_orient = 3 if is_free_ori else 1
+    A = np.zeros((G.shape[0], n_sources))
+    active_set = np.concatenate(-np.ones((n_sources, 1), dtype='int'))
+
+    G_proj = np.dot(np.identity(G.shape[0]), G)
+    phi_sig_proj = np.dot(np.identity(phi_sig.shape[0]), phi_sig)
+
+    for k in range(n_sources):
+        subcorr_max = -1
+        for i_source in range(G.shape[1] // n_orient):
+            if n_orient == 1:
+                Gk = G_proj[:, i_source]
+            else:
+                Gk = G_proj[:, n_orient * i_source:
+                            n_orient * i_source + n_orient]
+
+            subcorr, ori = _compute_subcorr(Gk, phi_sig_proj)
+            if subcorr > subcorr_max:
+                subcorr_max, active_set[k] = subcorr, i_source
+                A[:, k] = np.dot(Gk, ori)
+
+        logger.info("source %s found" % (k + 1))
+        if n_orient == 3:
+            logger.info("p = %s, ori = %s %s %s" % (active_set[k],
+                                                    ori[0], ori[1],
+                                                    ori[2]))
+
+        projection = _compute_proj(A[:, :k + 1])
+        G_proj = np.dot(projection, G)
```

Comment:
> On Feb 21, 2015, at 01:25, Alexandre Gramfort <notifications@github.com> wrote:
> 
> In mne/beamformer/_rap_music.py:
> 
> > +                Gk = G_proj[:, n_orient * i_source:
> > +                            n_orient * i_source + n_orient]
> > +
> > +            subcorr, ori = _compute_subcorr(Gk, phi_sig_proj)
> > +            if subcorr > subcorr_max:
> > +                subcorr_max, active_set[k] = subcorr, i_source
> > +                A[:, k] = np.dot(Gk, ori)
> > +
> > +        logger.info("source %s found" % (k + 1))
> > +        if n_orient == 3:
> > +            logger.info("p = %s, ori = %s %s %s" % (active_set[k],
> > +                                                    ori[0], ori[1],
> > +                                                    ori[2]))
> > +
> > +        projection = _compute_proj(A[:, :k + 1])
> > +        G_proj = np.dot(projection, G)
> G_proj = np.dot(projection, G_proj) ?
> 
We form the projection operator from the combination of A, not only the last vector. So we project the gain G and not G_proj

> —
> Reply to this email directly or view it on GitHub.
>